File: backend/src/modulos/auth/auth.servicio.js
// HU-01 Registrar cliente · HU-02 Iniciar y cerrar sesión · EN-03 Autenticación base
const bcrypt = require('bcryptjs');
const jwt = require('jsonwebtoken');
const config = require('../../config');
const { consultar } = require('../../db');
const { errores } = require('../../utils/errores');
const { ROLES } = require('../../middleware/autenticacion');

const RONDAS_BCRYPT = config.esPrueba ? 4 : 10;
const MENSAJE_CREDENCIALES = 'Correo o contraseña incorrectos'; // HU-02 criterio 2: mensaje genérico

// Hash de relleno: si el correo no existe igual comparamos, para que la
// respuesta tarde lo mismo y no se pueda adivinar qué correos existen.
const HASH_RELLENO = bcrypt.hashSync('contrasena-de-relleno', RONDAS_BCRYPT);

function usuarioPublico(fila) {
  return { id: fila.id, nombre: fila.nombre, correo: fila.correo, rol: fila.rol };
}

async function registrarCliente({ nombre, correo, contrasena }) {
  const existe = await consultar('SELECT 1 FROM usuarios WHERE correo = $1', [correo]);
  if (existe.rowCount > 0) {
    throw errores.conflicto('Ya existe una cuenta registrada con ese correo');
  }
  const hash = await bcrypt.hash(contrasena, RONDAS_BCRYPT);
  // El rol SIEMPRE es CLIENTE: aunque envíen "rol" en el cuerpo, se ignora (HU-01 criterio 5)
  const { rows } = await consultar(
    `INSERT INTO usuarios (nombre, correo, contrasena_hash, rol)
     VALUES ($1, $2, $3, $4)
     RETURNING id, nombre, correo, rol`,
    [nombre, correo, hash, ROLES.CLIENTE]
  );
  return usuarioPublico(rows[0]);
}

function minutosRestantes(fecha) {
  return Math.max(1, Math.ceil((fecha.getTime() - Date.now()) / 60000));
}

async function registrarIntentoFallido(usuarioId) {
  const { maxIntentosFallidos, minutosBloqueo } = config.login;
  // Si un bloqueo anterior ya venció, el contador vuelve a empezar desde 1
  const { rows } = await consultar(
    `UPDATE usuarios u
        SET intentos_fallidos = n.nuevo,
            bloqueado_hasta = CASE WHEN n.nuevo >= $2
                                   THEN now() + make_interval(mins => $3::int)
                                   ELSE NULL END
       FROM (SELECT id,
                    CASE WHEN bloqueado_hasta IS NOT NULL AND bloqueado_hasta <= now()
                         THEN 1 ELSE intentos_fallidos + 1 END AS nuevo
               FROM usuarios WHERE id = $1) n
      WHERE u.id = n.id
  RETURNING u.intentos_fallidos, u.bloqueado_hasta`,
    [usuarioId, maxIntentosFallidos, minutosBloqueo]
  );
  return rows[0];
}

async function iniciarSesion({ correo, contrasena, recordarme }, contexto = {}) {
  const { rows } = await consultar('SELECT * FROM usuarios WHERE correo = $1', [correo]);
  const usuario = rows[0];

  if (!usuario) {
    await bcrypt.compare(contrasena, HASH_RELLENO);
    throw errores.noAutenticado(MENSAJE_CREDENCIALES);
  }

  // HU-02 criterio 5: cuenta bloqueada temporalmente
  if (usuario.bloqueado_hasta && usuario.bloqueado_hasta.getTime() > Date.now()) {
    throw errores.bloqueado(
      `Cuenta bloqueada temporalmente por varios intentos fallidos. Intenta de nuevo en ${minutosRestantes(usuario.bloqueado_hasta)} minuto(s)`
    );
  }

  const correcta = await bcrypt.compare(contrasena, usuario.contrasena_hash);
  if (!correcta) {
    const estado = await registrarIntentoFallido(usuario.id);
    if (estado.bloqueado_hasta) {
      throw errores.bloqueado(
        `Cuenta bloqueada temporalmente por ${config.login.maxIntentosFallidos} intentos fallidos. Intenta de nuevo en ${config.login.minutosBloqueo} minutos`
      );
    }
    throw errores.noAutenticado(MENSAJE_CREDENCIALES);
  }

  // HU-15 criterio 2: un cliente desactivado no puede iniciar sesión
  if (!usuario.activo) {
    throw errores.prohibido('Tu cuenta está desactivada. Comunícate con el administrador');
  }

  await consultar(
    'UPDATE usuarios SET intentos_fallidos = 0, bloqueado_hasta = NULL WHERE id = $1',
    [usuario.id]
  );

  // HU-02 criterio 6: "Recordarme" extiende la sesión
  const s = config.sesion;
  const duracionMs = recordarme
    ? s.recordarmeDuracionDias * 24 * 60 * 60 * 1000
    : s.duracionHoras * 60 * 60 * 1000;
  const expiraEn = new Date(Date.now() + duracionMs);

  const sesion = await consultar(
    `INSERT INTO sesiones (usuario_id, recordarme, expira_en, ip, agente_usuario)
     VALUES ($1, $2, $3, $4, $5)
     RETURNING id`,
    [usuario.id, recordarme, expiraEn, contexto.ip || null, (contexto.agenteUsuario || '').slice(0, 255) || null]
  );

  const token = jwt.sign(
    { sid: sesion.rows[0].id, rol: usuario.rol },
    config.jwtSecreto,
    { subject: String(usuario.id), expiresIn: Math.floor(duracionMs / 1000) }
  );

  return {
    token,
    expiraEn: expiraEn.toISOString(),
    recordarme,
    usuario: usuarioPublico(usuario),
  };
}

/** HU-02 criterio 4: cerrar sesión invalida el token */
async function cerrarSesion(sesionId) {
  await consultar('UPDATE sesiones SET revocada_en = now() WHERE id = $1 AND revocada_en IS NULL', [sesionId]);
}

module.exports = { registrarCliente, iniciarSesion, cerrarSesion, MENSAJE_CREDENCIALES };

File: backend/src/main/java/alquiler/auth/AuthServicio.java
package alquiler.auth;

import alquiler.bd.Fila;
import alquiler.config.Config;
import alquiler.json.Json;
import alquiler.seguridad.Contrasenas;
import alquiler.seguridad.Jwt;
import alquiler.seguridad.Rol;
import alquiler.util.ErrorApp;

import java.time.Duration;
import java.time.Instant;
import java.util.Map;

/** HU-01 Registrar cliente · HU-02 Iniciar y cerrar sesión · EN-03 Autenticación base */
public class AuthServicio {

    /** HU-02 criterio 2: el mismo mensaje si el correo no existe o la contraseña es incorrecta. */
    public static final String MENSAJE_CREDENCIALES = "Correo o contraseña incorrectos";

    private final UsuarioRepositorio usuarios;
    private final SesionRepositorio sesiones;
    private final Contrasenas contrasenas;
    private final Jwt jwt;
    private final Config config;
    /** Hash de relleno: si el correo no existe igual se compara, para que la respuesta tarde lo mismo. */
    private final String hashRelleno;

    public AuthServicio(UsuarioRepositorio usuarios, SesionRepositorio sesiones,
                        Contrasenas contrasenas, Jwt jwt, Config config) {
        this.usuarios = usuarios;
        this.sesiones = sesiones;
        this.contrasenas = contrasenas;
        this.jwt = jwt;
        this.config = config;
        this.hashRelleno = contrasenas.cifrar("contrasena-de-relleno");
    }

    public static Map<String, Object> usuarioJson(Fila u) {
        return Json.obj("id", u.entero("id"), "nombre", u.texto("nombre"), "correo", u.texto("correo"), "rol", u.texto("rol"));
    }

    /** HU-01: el rol siempre es CLIENTE (criterio 5), aunque envíen otro. */
    public Map<String, Object> registrarCliente(String nombre, String correo, String contrasena) {
        if (usuarios.existeCorreo(correo)) {
            throw ErrorApp.conflicto("Ya existe una cuenta registrada con ese correo");
        }
        String hash = contrasenas.cifrar(contrasena);
        return usuarioJson(usuarios.crear(nombre, correo, hash, Rol.CLIENTE.name()));
    }

    public Map<String, Object> iniciarSesion(String correo, String contrasena, boolean recordarme,
                                             String ip, String agenteUsuario) {
        Fila u = usuarios.buscarPorCorreo(correo);
        if (u == null) {
            contrasenas.verificar(contrasena, hashRelleno);
            throw ErrorApp.noAutenticado(MENSAJE_CREDENCIALES);
        }

        // HU-02 criterio 5: cuenta bloqueada temporalmente
        Instant bloqueadoHasta = u.instante("bloqueado_hasta");
        if (bloqueadoHasta != null && bloqueadoHasta.isAfter(Instant.now())) {
            throw ErrorApp.bloqueado("Cuenta bloqueada temporalmente por varios intentos fallidos. Intenta de nuevo en "
                    + minutosRestantes(bloqueadoHasta) + " minuto(s)");
        }

        if (!contrasenas.verificar(contrasena, u.texto("contrasena_hash"))) {
            Fila estado = usuarios.registrarIntentoFallido(u.entero("id"), config.maxIntentosFallidos, config.minutosBloqueo);
            if (estado.instante("bloqueado_hasta") != null) {
                throw ErrorApp.bloqueado("Cuenta bloqueada temporalmente por " + config.maxIntentosFallidos
                        + " intentos fallidos. Intenta de nuevo en " + config.minutosBloqueo + " minutos");
            }
            throw ErrorApp.noAutenticado(MENSAJE_CREDENCIALES);
        }

        // HU-15 criterio 2: un cliente desactivado no puede iniciar sesión
        if (!u.bool("activo")) {
            throw ErrorApp.prohibido("Tu cuenta está desactivada. Comunícate con el administrador");
        }
        usuarios.reiniciarIntentos(u.entero("id"));

        // HU-02 criterio 6: "Recordarme" extiende la sesión
        Duration duracion = recordarme
                ? Duration.ofDays(config.recordarmeDuracionDias)
                : Duration.ofHours(config.duracionHoras);
        Instant expiraEn = Instant.now().plus(duracion);

        String agente = agenteUsuario == null ? null
                : agenteUsuario.substring(0, Math.min(255, agenteUsuario.length()));
        String sesionId = sesiones.crear(u.entero("id"), recordarme, expiraEn, ip, agente);
        String token = jwt.firmar(Json.obj("sid", sesionId, "rol", u.texto("rol")), String.valueOf(u.entero("id")), expiraEn);

        return Json.obj(
                "token", token,
                "expiraEn", expiraEn,
                "recordarme", recordarme,
                "usuario", usuarioJson(u));
    }

    /** HU-02 criterio 4 */
    public void cerrarSesion(String sesionId) {
        sesiones.revocar(sesionId);
    }

    private static long minutosRestantes(Instant hasta) {
        long segundos = Duration.between(Instant.now(), hasta).getSeconds();
        return Math.max(1, (segundos + 59) / 60);
    }
}
